### ISSUE TITLE
[Build fails] multiple definition of `renderer_sync'

### ISSUE THREAD

**Issue by SupervisedThinking** [libretro/pcsx_rearmed#494]
## Description

if you build pcsx_rearmed for ARM devices with `BUILTIN_GPU=neon` enabled `THREAD_RENDERING=1` you end up with `multiple definition of `renderer_sync'` and else see log:

https://pastebin.com/feVUaeLW

## Steps to reproduce

package:
https://github.com/SupervisedThinking/LibreELEC-RR/blob/master-rr/packages/supervisedthinking/emulation/libretro/pcsx_rearmed/package.mk

## When did the behavior start?

I'm pretty sure it's related to https://github.com/libretro/pcsx_rearmed/pull/462 which was only tested for GPU=unai
## Your device/OS/platform/architecture

LibreELEC / ARM neon builds

**Comment by negativeExponent** [comment]
THREADED_RENDERING probably wont even work even if you fix it, since it was not tested on neon.

you can probably apply similar function replacement done on unai and see if it does work on neon:

https://github.com/justinweiss/pcsx_rearmed/blob/c765eb86debdc06fe304511bc2edbb6f3e3d7813/plugins/gpu_unai/gpulib_if.cpp#L29-L43

**Comment by SupervisedThinking** [comment]
I'm aware that it's "unai only" and so I was wondering why this was added in the first place if nobody tests it for all platforms: 

https://github.com/libretro/pcsx_rearmed/blob/master/Makefile.libretro#L51

https://github.com/libretro/pcsx_rearmed/commit/54b375e0d7031d0fae9deb0f7ba815793ae4115e#diff-9b06322f74c0fac1f20a02120275946e3e2ec8e150ef3edfd424f8ac1c345799

**Comment by negativeExponent** [comment]
threading is not enabled on default. why make a fuzz about it...

**Comment by SupervisedThinking** [comment]
Well what? https://github.com/libretro/pcsx_rearmed/blob/master/Makefile.libretro#L47-L53

```
# Unix
ifeq ($(platform), unix)
	TARGET := $(TARGET_NAME)_libretro.so
	fpic := -fPIC
	THREAD_RENDERING = 1
ifneq ($(findstring SunOS,$(shell uname -s)),)
	CC = gcc
```

`THREAD_RENDERING = 1` does mean it's not enabled for Linux as default? Ah i see...

**Comment by negativeExponent** [comment]
if you build with "neon" then disable threading. its a very simple thing. no need to be sarcastic in your reply. ill wait for the the time when you do "make" something better than this!

**Comment by justinweiss** [comment]
I'm surprised this hasn't come up before -- it seems like most SBCs would use linux+neon, and this change has been around for months. Maybe the buildbot doesn't cover it.

Do you have a way to build and test this core? If so, I think adding a block like like this to `gpu_neon/psx_gpu_if.c` should get it working: https://github.com/libretro/pcsx_rearmed/pull/464/files#diff-ec2bb7beb4cd92a15bd9c9274aa51eaed75595b905bd2cbad201221cd63a4b20R20-R35. It would be worth a try to see if it improves the speed.

If not, always disabling threaded rendering when neon is enabled seems best.

**Comment by SupervisedThinking** [comment]
@negativeExponent normally I would just insert a random _facepalm meme_ to answer you appropriate but I try to act like an adult.
From my perspective I've delivered everything needed to understand what the problem is, I've digged through the commits to find the "bad" patchset & pointed out that `THREAD_RENDERING=1` is broken for `BUILTIN_GPU=neon` which is basically the standard more or less for **all SBCs*** around unless you use something like a RPi1 with a BCM2835 SoC. I've also pointed out that the makefile enables `THREAD_RENDERING` for ***all*** Unix-like systems which is IMHO just wrong.

Instead of admitting that you either haven't read my links or understood the problem you still try to convince me that it's the best way to overwrite false flags instead of fixing the makefile. As I said... insert a random _facepalm meme_

@justinweiss  it came up when I rebased my stuff for LE10 & also the upstream Kodi Game package was broken after a recent bump so no clue why other builds did not ran into problems since linux+neon is, as you've mentioned, pretty much the standard for SBCs.

Sure if you've some patches or a wip branch I can ran some build tests & run test it on RK3399 & A311D SoCs.

The core itself builds fine for generic x86_64 builds so how about something like:

```
ifeq ($(ARCH), $(filter $(ARCH), arm))
   THREAD_RENDERING = 0
else
   THREAD_RENDERING = 1
endif
```

And then manually enable it for those few ARM SBCs which lack NEON? Well unless you'll fix the build for gpus beside unai.

**Comment by negativeExponent** [comment]
your crazy. you stated "neon", then you linked about the section on the makefile about "linux" platform having threading enabled by default, when what i said about threading disabled by default is on other "known" platforms other than linux and ctr. then you linked a so called "fixed" on your end which targets specifically "arm" platform, how is this related? i dont even know what shitty device this is

again i have told you above threading was not tested on neon, so it should not be enabled. even linked a section to try if the threading works on neon as @justinweiss, the one who added threading support has not tested neon, hence its not suppose to be enabled, if it did enable itself, then there is a flag to disable. again there was no need to escalate the situation, but you decided to sarcastic about your reply in https://github.com/libretro/pcsx_rearmed/issues/494#issuecomment-775227381, specifically when and what words are indicated in **bold**.

again there was no need to even start escalating it, but you started it.. so lets continue...

**Comment by SupervisedThinking** [comment]
@negativeExponent 
_tl:dr almost if not all SoC for SBC support NEON & the vast majority of them runs Linux which results in failed builds because `THREAD_RENDERING`, which is now enabled at default for Linux, was only tested for `BUILTIN_GPU = unai` which is only used by few low-end SoC or non 32bit userspace builds._

Please do yourself a favor and ***delete all of your comments*** because they show a serious lack of knowledge about ARM SoCs, SBCs, Linux and well basically everything this is related to. Unless you use an ancient cheap ass crap board dropped by a truck you won't easily find any somewhat recent SoC which does not support NEON. So if you only test threaded rendering for `BUILTIN_GPU = unai` then it should only be enabled per default for platforms that actually use `BUILTIN_GPU = unai` like `CTR(3DS)`, `libnx` or `rpi4_64`. But it wasn't. It was & is enabled for everything Unix-like no matter which CPU or SoC it uses & if it'll support NEON or not. If you don't even know what _shitty device_ `ARCH=arm` is then I recommend rtfm:

https://github.com/libretro/Lakka-LibreELEC/blob/master/projects/Amlogic/devices/AMLGX/options#L13
https://github.com/libretro/Lakka-LibreELEC/blob/master/projects/RPi/options#L5-L8
https://github.com/libretro/Lakka-LibreELEC/blob/master/projects/OdroidXU3/options#L8
https://github.com/libretro/Lakka-LibreELEC/blob/master/projects/Qualcomm/devices/Dragonboard/options#L13
https://github.com/libretro/Lakka-LibreELEC/blob/master/projects/Rockchip/devices/RK3399/options#L13

My patch enables `THREAD_RENDERING=1` for everything which is not `ARCH=arm` or `ARCH=aarch64`, the logic could be reversed and `ARCH=Generic` used as condition, doesn't matter in the end. 
For everything else like 64bit SBC builds, 3DS or switch `THREAD_RENDERING=1` could be manually set & added to the platform specs ***once it was tested*** and most likely everything will be fine 🌈 🦄 

It's fine that you're probably a x86 Windows or macOS user which isn't able to understand how those things in particular are interacting but then again just don't comment. Thank you.

**Comment by justinweiss** [comment]
Here's two patches to try out.

This is an attempt to support threaded rendering for neon. It adds a new core option for Threaded Rendering: off, sync, and async. It might help lower-powered devices, maybe fast-forward -- not sure whether the neon plugin is ever limited by rendering speed.

[0001-Enable-threaded-rendering-for-neon.patch.txt](https://github.com/libretro/pcsx_rearmed/files/5962889/0001-Enable-threaded-rendering-for-neon.patch.txt)

If that doesn't work, this is a patch to force-disable threaded rendering when gpu_neon is enabled:

[0001-Force-disable-threaded-rendering-for-neon.patch.txt](https://github.com/libretro/pcsx_rearmed/files/5962894/0001-Force-disable-threaded-rendering-for-neon.patch.txt)

`git am` should be able to apply these to the current master branch.

**Comment by SupervisedThinking** [comment]
@justinweiss this patch compiles fine, haven't runtested yet:
```
--- a/Makefile
+++ b/Makefile
@@ -194,9 +194,11 @@ plugins/gpu_unai/gpulib_if.o: CFLAGS += -DREARMED -O3
 CC_LINK = $(CXX)
 endif
 ifeq "$(THREAD_RENDERING)" "1"
+ifneq "$(BUILTIN_GPU)" "neon"
 CFLAGS += -DTHREAD_RENDERING
 OBJS += plugins/gpulib/gpulib_thread_if.o
 endif
+endif
```

The other one fails:
```
/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RPi4.arm-9.80-devel/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc -march=armv8-a+crc -mtune=cortex-a53 -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mfloat-abi=hard -mfpu=neon-fp-armv8 -Wall -pipe  -O2 -fomit-frame-pointer -DGIT_VERSION=\"976907a\" -marm -fPIC -D_FILE_OFFSET_BITS=64 -Wall -Iinclude -ffast-math -O2 -DNDEBUG -Ideps/zlib -DNEW_DYNAREC -DGPU_NEON -DTHREAD_RENDERING -Ideps/libchdr -DHAVE_SYS_PARAM_H -Ideps/crypto -Ideps/flac-1.3.2/include -Ideps/flac-1.3.2/src/libFLAC/include -Ideps/flac-1.3.2/src/libFLAC/include -Ideps/lzma-16.04/C -DHAVE_CHD -D'PACKAGE_VERSION="1.3.2"' -DFLAC__HAS_OGG=0 -DFLAC__NO_DLL -DHAVE_LROUND -DHAVE_STDINT_H -DHAVE_STDLIB_H -DFLAC__NO_DLL -D_7ZIP_ST -Ilibretro-common/include -DFRONTEND_SUPPORTS_RGB565 -DHAVE_LIBRETRO -DNO_FRONTEND -DNEON_BUILD -DTEXTURE_CACHE_4BPP -DTEXTURE_CACHE_8BPP -c -o plugins/gpu_neon/psx_gpu_if.o plugins/gpu_neon/psx_gpu_if.c
plugins/gpu_neon/psx_gpu_if.c: In function 'real_do_cmd_list':
plugins/gpu_neon/psx_gpu_if.c:46:17: error: 'scratch_ex_regs' undeclared (first use in this function)
   46 | #define ex_regs scratch_ex_regs
      |                 ^~~~~~~~~~~~~~~
plugins/gpu_neon/psx_gpu_if.c:60:3: note: in expansion of macro 'ex_regs'
   60 |   ex_regs[1] &= ~0x1ff;
      |   ^~~~~~~
plugins/gpu_neon/psx_gpu_if.c:46:17: note: each undeclared identifier is reported only once for each function it appears in
   46 | #define ex_regs scratch_ex_regs
      |                 ^~~~~~~~~~~~~~~
plugins/gpu_neon/psx_gpu_if.c:60:3: note: in expansion of macro 'ex_regs'
   60 |   ex_regs[1] &= ~0x1ff;
      |   ^~~~~~~
plugins/gpu_neon/psx_gpu_if.c: In function 'real_renderer_init':
plugins/gpu_neon/psx_gpu_if.c:46:17: error: 'scratch_ex_regs' undeclared (first use in this function)
   46 | #define ex_regs scratch_ex_regs
      |                 ^~~~~~~~~~~~~~~
plugins/gpu_neon/psx_gpu_if.c:106:3: note: in expansion of macro 'ex_regs'
  106 |   ex_regs = gpu.ex_regs;
      |   ^~~~~~~
make[1]: *** [Makefile:383: plugins/gpu_neon/psx_gpu_if.o] Error 1
```

Anyway the first patch "fakes threaded" rendering but opt it out in the `Makefile` for `GPU=neon` so people will assume it's enabled if they have a look in the `makefile.libretro` then? I'm not sure if this is a proper solution 🤔

**Comment by justinweiss** [comment]
Ah, it looks like gpu_neon has its own version of ex_regs -- moving the THREAD_RENDERING block above that declaration above that should fix the problem:

[0001-Enable-threaded-rendering-for-neon.patch.txt](https://github.com/libretro/pcsx_rearmed/files/5967013/0001-Enable-threaded-rendering-for-neon.patch.txt)

Is there an easy way to set up a cross-compiler for this? If so, I can try myself instead of shipping patches back and forth.

> Anyway the first patch "fakes threaded" rendering but opt it out in the Makefile for GPU=neon so people will assume it's enabled if they have a look in the makefile.libretro then?

More like "The threaded rendering variable only has an effect for non-neon gpu plugins." It's the same way it used to work when it was only implemented for unai -- "The variable is only checked when the plugin is unai."

**Comment by SupervisedThinking** [comment]
Latest patch:

```
/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-9.80-devel/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc -march=armv8-a+crc -mtune=cortex-a72.cortex-a53 -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mfloat-abi=hard -mfpu=crypto-neon-fp-armv8 -Wall -pipe  -O2 -fomit-frame-pointer -DGIT_VERSION=\"976907a\" -marm -fPIC -D_FILE_OFFSET_BITS=64 -Wall -Iinclude -ffast-math -O2 -DNDEBUG -Ideps/zlib -DNEW_DYNAREC -DGPU_NEON -DTHREAD_RENDERING -Ideps/libchdr -DHAVE_SYS_PARAM_H -Ideps/crypto -Ideps/flac-1.3.2/include -Ideps/flac-1.3.2/src/libFLAC/include -Ideps/flac-1.3.2/src/libFLAC/include -Ideps/lzma-16.04/C -DHAVE_CHD -D'PACKAGE_VERSION="1.3.2"' -DFLAC__HAS_OGG=0 -DFLAC__NO_DLL -DHAVE_LROUND -DHAVE_STDINT_H -DHAVE_STDLIB_H -DFLAC__NO_DLL -D_7ZIP_ST -Ilibretro-common/include -DFRONTEND_SUPPORTS_RGB565 -DHAVE_LIBRETRO -DNO_FRONTEND -c -o frontend/plugin.o frontend/plugin.c
In file included from plugins/gpu_neon/psx_gpu_if.c:20:
plugins/gpu_neon/../gpulib/gpulib_thread_if.h:27:23: error: unknown type name 'uint32_t'
   27 | int  real_do_cmd_list(uint32_t *list, int count, int *last_cmd);
      |                       ^~~~~~~~
plugins/gpu_neon/../gpulib/gpulib_thread_if.h:30:31: error: unknown type name 'uint32_t'
   30 | void real_renderer_sync_ecmds(uint32_t * ecmds);
      |                               ^~~~~~~~
plugins/gpu_neon/../gpulib/gpulib_thread_if.h:34:44: warning: 'struct rearmed_cbs' declared inside parameter list will not be visible outside of this definition or declaration
   34 | void real_renderer_set_config(const struct rearmed_cbs *config);
      |                                            ^~~~~~~~~~~
In file included from plugins/gpu_neon/psx_gpu_if.c:47:
plugins/gpu_neon/../gpulib/gpu.h:96:12: error: duplicate member 'scratch_ex_regs'
   96 |   uint32_t scratch_ex_regs[8]; // for threaded rendering
      |            ^~~~~~~~~~~~~~~
plugins/gpu_neon/psx_gpu_if.c:28:29: error: conflicting types for 'real_renderer_set_config'
   28 | #define renderer_set_config real_renderer_set_config
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~
plugins/gpu_neon/../gpulib/gpu.h:120:6: note: in expansion of macro 'renderer_set_config'
  120 | void renderer_set_config(const struct rearmed_cbs *config);
      |      ^~~~~~~~~~~~~~~~~~~
In file included from plugins/gpu_neon/psx_gpu_if.c:20:
plugins/gpu_neon/../gpulib/gpulib_thread_if.h:34:6: note: previous declaration of 'real_renderer_set_config' was here
   34 | void real_renderer_set_config(const struct rearmed_cbs *config);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~
plugins/gpu_neon/psx_gpu_if.c:28:29: error: conflicting types for 'real_renderer_set_config'
   28 | #define renderer_set_config real_renderer_set_config
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~
plugins/gpu_neon/psx_gpu_if.c:187:6: note: in expansion of macro 'renderer_set_config'
  187 | void renderer_set_config(const struct rearmed_cbs *cbs)
      |      ^~~~~~~~~~~~~~~~~~~
In file included from plugins/gpu_neon/psx_gpu_if.c:20:
plugins/gpu_neon/../gpulib/gpulib_thread_if.h:34:6: note: previous declaration of 'real_renderer_set_config' was here
   34 | void real_renderer_set_config(const struct rearmed_cbs *config);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~
make: *** [Makefile:383: plugins/gpu_neon/psx_gpu_if.o] Error 1
```

Technically you could clone my repo: https://github.com/SupervisedThinking/LibreELEC-RR and run:

`PROJECT=Rockchip DEVICE=RK3399 ARCH=arm scripts/build pcsx_rearmed`

**Comment by justinweiss** [comment]
Thanks, I was able to use that to build the toolchain, and get this revised version of the patch to build successfully, if you'd like to give it a try:

[0001-Enable-threaded-rendering-for-neon.patch.txt](https://github.com/libretro/pcsx_rearmed/files/5970055/0001-Enable-threaded-rendering-for-neon.patch.txt)

Again, this adds a "Threaded Rendering" core setting that, when enabled, will run GPU commands in a background thread.

BTW, I agree that your patch makes it more obvious that the setting is disabled. My approach is that `THREAD_RENDERING` + `GPU_NEON` will never work without a patch to support it. So if some other device comes along that is linux + neon but somehow avoids the other clauses disabling `THREAD_RENDERING`, it still shouldn't break -- it's just that `THREAD_RENDERING` does nothing, because that combination of settings doesn't make sense. If they ask for both, they're only going to get one, or they'll get an error. It would be like if it we had something like `UNAI_THREAD_RENDERING=1` along with `BUILTIN_GPU=GPU_NEON`, one of them is going to get ignored, even if it's just implicitly through how the source treats them.

So I have no problem explicitly disabling `THREAD_RENDERING` in those situations if that makes it more obvious, but I also think it should be automatically disabled if neon is enabled (without this patch to support it), because it doesn't apply.

**Comment by SupervisedThinking** [comment]
In an ideal world we had a build system that detects target cpu features & adds features as they are available. I'm aware that threaded rendering adds an additional setting to core setting, it's also present in my core for generic builds. Of course I have no clue if it works for generic systems but since it does neither unai nor neon as gpu it's likely.

IMHO the best way would to test which systems work with threaded rendering and add this flag as platform opt & get rid of it otherwhise. From what I've found in the makefile I guess the neon gpu is 32bit only? This would explain why 64bit builds use unai too which was IIRC added for stuff like 3DS which is really low end compared to everything else.

**Comment by justinweiss** [comment]
With that latest patch, did threaded rendering work on neon devices? The best option would be to support it everywhere, or at least everywhere we have threads -- I just don't have the hardware or toolchain to test.

**Comment by dankcushions** [comment]
Thanks @justinweiss - tried your patch and it builds, but crashes when set to 'sync' for rpi4 in tekken 3, just after I select a game mode. Crashes just after this line in log:
```
[INFO] SET_GEOMETRY: 368x480, aspect: 1.333.
```
weirdly when i in debug mode and via gdb i get an earlier crash:
```
Thread 1 "retroarch" received signal SIGSEGV, Segmentation fault.
shade_blocks_textured_unmodulated_direct () at plugins/gpu_neon/psx_gpu/psx_gpu_arm_neon.S:3430
warning: Source file is more recent than executable.
3430	  vld1.u16 { fb_pixels_next }, [fb_ptr_next]
(gdb) bt
#0  0xae1ea214 in shade_blocks_textured_unmodulated_direct () at plugins/gpu_neon/psx_gpu/psx_gpu_arm_neon.S:3430
#1  0xae1b6094 in flush_render_block_buffer (psx_gpu=0xae5f9700 <egpu>) at plugins/gpu_neon/psx_gpu/psx_gpu.c:505
#2  0xae1e6554 in real_renderer_flush_queues () at plugins/gpu_neon/psx_gpu_if.c:164
#3  0xae1fa2c8 in renderer_flush_queues () at plugins/gpulib/gpulib_thread_if.c:397
#4  0xae1b4b10 in GPUupdateLace () at plugins/gpulib/gpu.c:750
#5  0xae115370 in psxRcntUpdate () at libpcsxcore/psxcounters.c:340
#6  0xae149cd0 in irq_test () at libpcsxcore/new_dynarec/backends/psx/emu_if.c:86
#7  0xae149dd4 in gen_interupt () at libpcsxcore/new_dynarec/backends/psx/emu_if.c:101
#8  0xae1a2f60 in cc_interrupt () at libpcsxcore/new_dynarec/arm/linkage_arm.S:452
```
tried it a few more times and got a different one:
```
Thread 4 "retroarch" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xad42af30 (LWP 985)]
video_thread_main (arg=0xae9ff108 <thread>) at plugins/gpulib/gpulib_thread_if.c:81
warning: Source file is more recent than executable.
81			while (!thread->queue->used && thread->running) {
(gdb) bt
#0  0xae1f91f8 in video_thread_main (arg=0xae9ff108 <thread>) at plugins/gpulib/gpulib_thread_if.c:81
#1  0xb6e8b494 in start_thread (arg=0xad42af30) at pthread_create.c:486
#2  0xb5137578 in  () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
```
not quite sure what's going on but presume it's thread-related.

**Comment by justinweiss** [comment]
That's too bad. I was hoping it would be as simple as peops, but it looks like it will take more work. Probably best to only allow it for peops/unai builds for now. I'll put together a PR.
